main.go

```diff
@@ -73,11 +74,36 @@ func main() {
 		sigint := make(chan os.Signal, 1)
 		signal.Notify(sigint, os.Interrupt, syscall.SIGTERM)
 		<-sigint
-		s.stop <- struct{}{} // notify having caught signal
+		s.stop <- struct{}{}             // notify having caught signal stop oauthproxy
+		metricsStopChannel <- struct{}{} // and the metrics endpoint
 	}()
 	s.ListenAndServe()
 }
 
+// startMetricsServer will start the metrics server on the specified address.
+// It always return a channel to signal stop even when it does not run.
+func startMetricsServer(address string) chan struct{} {
+	stop := make(chan struct{}, 1)
+
+	// Attempt to setup the metrics endpoint if we have an address
+	if address != "" {
+		s := &http.Server{Addr: address, Handler: middleware.DefaultMetricsHandler}
+		go s.ListenAndServe()
+
+		go func() {
+			<-stop
+			if err := s.Shutdown(context.Background()); err != nil {
+				logger.Fatal(err)
+			}
+		}()
+	} else { // Capture the stop signal as not block shutdown
+		go func() {
+			<-stop
+		}()
```

Comment:
I wonder if instead of writing to these channels, we need to close them instead? That would mean we don't need to have this block. Reading from a closed channel doesn't block so works in the same way as writing to a channel, but you don't have anything read from it

Reply:
I've removed this and now close the channel instead. The metrics HTTP server still closes so I think this was redundant. Thanks for catching that.